### ISSUE TITLE
word and byte overflow and carry bits

### ISSUE THREAD

**Issue by Zardoz89** [trillek-team/trillek-computer#41]
Related with issue #38 

Actually byte and word size math operations (add, sub, rsb) not have an carry or overflow flags. Carry flag for word and byte, not is highly necesary, as these bit could be extracted/compared on the result register, but the overflow condition is more complex. Doing it by software could carry an unfair penaly to word and byte wide math operations.

I propose add to %flag, this flag bits : 

- OBF : Overflow Byte Flag - Overflow condition on a 8 bit wide math operation
- OWF : Overflow Word Flag - Overflow condition on a 16 bit wide math operation.

To check the condition, would be like on normal overflow flag, but using the 7th or 15th bit as sign bit.
```
       ADDITION SIGN BITS           SUBTRACTION SIGN BITS              
    num1sign num2sign sumsign    num1sign num2sign sumsign                
    ---------------------------  ---------------------------
     0       0        1           0        1        1
     1       1        0           1        0        0
```
For example, for addition : OF = SIGN_BIT(num1) == DW_SIGN_BIT(num2) && DW_SIGN_BIT(num1) != DW_SIGN_BIT(result)

Optionally, and by concordance , we could add : 

- CBF : Carry Byte Flag - Carry byte of a 8 bit wide math operation. Same value that 8th bit of result
- CWF : Carry Word Flag - Carry byte of a 16 bit wide math operation. Same value that 16th bit of result.

**Comment by milesrout** [comment]
It seems reasonable to include at all operand sizes both arithemetical functions that record carry/overflow in some way and functions that do not, if doing so has a performance benefit. If it provides no benefit over some sort of toggle of 'record carry/overflow mode', or over just tracking carry/overflow in software then we shouldn't add the unnecessary complexity.

**Comment by alexfru** [comment]
There are two common ways to set the carry and overflow flags following addition/subtraction and it looks like you can get carry/overflow flags for 8-bit and 16-bit add/sub cheaply. E.g. [Overflow and Carry flags on Z80](http://stackoverflow.com/questions/8034566/overflow-and-carry-flags-on-z80).

**Comment by milesrout** [comment]
If you add n-bit numbers as (2n)-bit numbers, you can get the carry/overflow flags pretty trivially.

**Comment by alexfru** [comment]
Right, I'm saying it's easy and cheap to provide multiple CY/OV flags for different positions within the machine word if you add two machine words.

**Comment by milesrout** [comment]
I have no idea what you just said.

**Comment by alexfru** [comment]
The implementation is on stack overflow in my answer, see the link.

**Comment by milesrout** [comment]
If what you mean is 'To get 8-bit addition with carry, sign-extend them (if they are signed) or zero-extend them (if they are unsigned) to 16-bits, do the addition in 16-bits, then truncate for the result and look at the 9th bit of the 16-bit word for the carry bit' then yes, that's exactly what we should be doing. And the same for 16-bit/32-bit and 32-bit/64-bit.

**Comment by alexfru** [comment]
I mean that in hardware you get carries in positions 7/8, 15/16, 31/32 essentially for free when you do 32-bit add/sub. Signed overflows are just XORed carries in positions 7 and 8, 15 and 16 and 31 and 32. In software (in a CPU simulator) it's also easy to get the carries in C (one of the two methods described in my answer on stackoverflow (between \#else and \#endif), based on XORing the result with the input operands) does that.

**Comment by milesrout** [comment]
Okay well if we can do it cheaply then we might as well just do it for all arithmetic, rather than having arithmetic instructions that carry and ones that don't.

**Comment by alexfru** [comment]
It might not be as straightforward with multiplication and division, though. Except, perhaps, if you analyze the product/quotient/remainder for fitting into 8/16/32 bits (the most significant bits of the result being all zeroes or sign bits).

**Comment by milesrout** [comment]
Just sign-extend everything to 64-bit, do normal addition and check the 9th/17th/33rd bit.

**Comment by Zardoz89** [comment]
Multiplication/division actually cleans overflow and carry flags . Should have other behaviour ?

The actual implementation, does carry and overflow, with "if's" checking operands and the result. I'm reading your method. Indeed, very clever and could do the job very quick avoiding branching. Do you have a example of getting the overflow ?

**Comment by milesrout** [comment]
You can do overflow/carry checking without branches.

```
u8 a, b;  // operands
u8 c;     // carry-in
u8 C = ((u16)a + (u16)b + c) >> 8;  // carry-out
u8 O = ~((a >> 7) ^ (b >> 7)) & ((a >> 7) ^ ((a+b+c) >> 7)); // overflow
```

**Comment by Zardoz89** [comment]
Take from a verilog 6502 ALU : `V = AI7 ^ BI7 ^ CO ^ N;` (N is Negative flag; CO is carry)
Also, I noticed that Z80 and 6502 verilog cores always sets overflow flag with this condition. This points me that I put a strange description of overflow flag on the specs, that should be fixed.

**Comment by milesrout** [comment]
Reminder: 

the carry flag indicates whether the operation overflowed when considered as an unsigned operation.
the overflow flag indicates whether the operation overflowed when considered as a signed operation.
